### DISCUSSION TITLE
No Connection to Viessmann AP with Linux

### DISCUSSION THREAD

**Discussion by Nanadoe** [open3e/open3e#96] [Q&A]
Hi all,

First of all, i wanted to thank all of you involved in the Open3E project.  It has already been a great help to understand the configuration of our Viessmann system and to make suggestions to the Viessmann technician who had to come a few times already to modify the settings.

I have a Vitocal 250 / VX3 and E380.  I used Open3E successfully in Windows to read out the datapoints of the Heatpump.
I now wanted to get also data about the VX3 and Open3E, the data in the vicare app are not complete/correct.  If i understand it correct, you need Open3E on Linux for that. It is over 20 years I last used Linux, it was a good occassion to start again.

I installed Debian on an old laptop, as the Wifi adapter is broken, i'm using a dongle.  That all works fine, but when I want to connect the Wifi to the Viessmann Access Point, DHCP seems to fail. Do you need some specific configuration to access the Viessmann AP under Linux?

Here is the configuration of the connection in /etc/NetworkManager/system-connections/VIESSMANN-xxxx.nmconnection 

[connection]
id=VIESSMANN-xxxx
uuid=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
type=wifi
autoconnect=false
interface-name=wlx002df45b41e0

[wifi]
mode=infrastructure
ssid=VIESSMANN-xxxx

[wifi-security]
key-mgmt=wpa-psk
psk=xxxxxxxxxxxxxxxxxxx

[ipv4]
method=auto

[ipv6]
addr-gen-mode=default
method=auto
[proxy]

i get the following output from journalctl -xe NM_CONNECTION=xxxxxxxxxxxxxxxxx+ NM_DEVICE=wlx002df45b41e0'

pr 21 21:18:35 debian NetworkManager[729]: <info>  [1713727115.8221] device (wlx002df45b41e0): supplicant interface state: inactive -> disconnected
apr 21 21:18:35 debian NetworkManager[729]: <info>  [1713727115.8318] device (wlx002df45b41e0): supplicant interface state: disconnected -> inactive
apr 21 21:18:35 debian NetworkManager[729]: <info>  [1713727115.8616] device (wlx002df45b41e0): supplicant interface state: inactive -> authenticating
apr 21 21:18:35 debian NetworkManager[729]: <info>  [1713727115.9107] device (wlx002df45b41e0): supplicant interface state: authenticating -> associating
apr 21 21:18:35 debian NetworkManager[729]: <info>  [1713727115.9329] device (wlx002df45b41e0): supplicant interface state: associating -> associated
apr 21 21:18:36 debian NetworkManager[729]: <info>  [1713727116.1369] device (wlx002df45b41e0): supplicant interface state: associated -> 4way_handshake
apr 21 21:18:36 debian NetworkManager[729]: <info>  [1713727116.1883] device (wlx002df45b41e0): supplicant interface state: 4way_handshake -> completed
apr 21 21:18:36 debian NetworkManager[729]: <info>  [1713727116.1884] device (wlx002df45b41e0): Activation: (wifi) Stage 2 of 5 (Device Configure) successful. Connected to wireless network "VIESSMANN-xxxx"
apr 21 21:18:36 debian NetworkManager[729]: <info>  [1713727116.1885] device (wlx002df45b41e0): state change: config -> ip-config (reason 'none', sys-iface-state: 'managed'
apr 21 21:18:36 debian NetworkManager[729]: <info>  [1713727116.1892] dhcp4 (wlx002df45b41e0): activation: beginning transaction (timeout in 45 seconds)
apr 21 21:19:20 debian NetworkManager[729]: <info>  [1713727160.9486] device (wlx002df45b41e0): state change: ip-config -> failed (reason 'ip-config-unavailable', sys-iface-state: 'managed')
apr 21 21:19:21 debian NetworkManager[729]: <info>  [1713727161.0013] device (wlx002df45b41e0): set-hw-addr: set MAC address to 76:86:A9:49:EF:19 (scanning)
apr 21 21:19:21 debian NetworkManager[729]: <warn>  [1713727161.0364] device (wlx002df45b41e0): Activation: failed for connection 'VIESSMANN-xxxx'
apr 21 21:19:21 debian NetworkManager[729]: <info>  [1713727161.0371] device (wlx002df45b41e0): state change: failed -> disconnected (reason 'none', sys-iface-state: 'managed')
apr 21 21:19:21 debian NetworkManager[729]: <info>  [1713727161.0571] dhcp4 (wlx002df45b41e0): canceled DHCP transaction
apr 21 21:19:21 debian NetworkManager[729]: <info>  [1713727161.0571] dhcp4 (wlx002df45b41e0): activation: beginning transaction (timeout in 45 seconds)
apr 21 21:19:21 debian NetworkManager[729]: <info>  [1713727161.0572] dhcp4 (wlx002df45b41e0): state changed no lease
apr 21 21:19:21 debian NetworkManager[729]: <info>  [1713727161.0587] device (wlx002df45b41e0): supplicant interface state: completed -> disconnected
apr 21 21:19:24 debian NetworkManager[729]: <info>  [1713727164.0874] device (wlx002df45b41e0): supplicant interface state: disconnected -> inactive

I also tried to use a fixed IP,  then a connection is established, but I do not see any other devices but my own laptop.
I'm running out of inspiration, and i was hoping someone could help me here. 

Many thanks for suggestions or tips! 

kind regards,
Nanadoe

**Comment by philippoo66** [comment]
Hi @Nanadoe !

if open3e is working on Windows for Vitocal there's no need to change to another system to 'talk' with VX3.

even I wonder how you managed to do the CAN hardware implementation on Windows, making it working with the stuff we use in open3e. It would be nice if you coud let us know... {edit: ok AP mode - no CAN required...}

With the network issue I think there are others who know much better than me. BUT what I can tell is that I recently tried to start working with debian on an 'old' laptop, and it was horrible! After a day of poking in the mud I installed Mint and everything was working out of the box within minutes. Ubuntu may be an alternative more 'in between' Debian and Mint, but I'm really satified with my choice. 

greetings!
Phil

**Reply by Nanadoe** [reply]
Hi @philippoo66 ,

Thank you for your answer!
Indeed, i used the AP method in windows, I suppose when the AP is directly integrated in the Vitocal, no can required

In windows, I ran python3 Open3Eclient.py -d 192.168.0.1 -v -a -dev vcal > blabla
which returns the output of all the DiDs to a file 'blabla'
Someone at the Viessmann forum suggested to use the IP address 192.168.0.1 but i can't find back that reference.

Running python3 Open3Eclient.py -d 192.168.0.1 -v -a -dev vx3 obviously fails:

socket.CMSG_SPACE not available on this platform
reading 0x680, 189 datapoints, please be patient...
Traceback (most recent call last):
  File "C:\Users\lauwe\open3e\Open3Eclient.py", line 400, in <module>
    lst = ecu.readAll(args.raw)
          ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lauwe\open3e\Open3Eclass.py", line 155, in readAll
    value,idstr = self.readByDid(int(did), raw=raw)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lauwe\open3e\Open3Eclass.py", line 139, in readByDid
    response = self.uds_client.read_data_by_identifier([did])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lauwe\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\udsoncan\client.py", line 167, in decorated
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lauwe\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\udsoncan\client.py", line 466, in read_data_by_identifier
    response = self.send_request(req)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\lauwe\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\udsoncan\client.py", line 2255, in send_request
    raise NegativeResponseException(response)
udsoncan.exceptions.NegativeResponseException: ReadDataByIdentifier service execution returned a negative response ConditionsNotCorrect (0x22)

Thank you for the suggestion to try a more modern distro :)

greetings,
Nanadoe

**Reply by philippoo66** [reply]
there seem to be a certain datapoint in the list wich is not supported to get read from your device. the did list for vx3 is 'general'. because of this we introduced open3e_depictsystem. but unfortunately this is not too stable... 

solution to your issue is to find out which did/s are not supported and to comment them out. do you know a little how to debug python (breakpoints, ad print(), ...)?

it will NOT help to run open3e on Linux in this case.

**Reply by Nanadoe** [reply]
hi all,
i've checked which datapoints in the list for vx3 are supported for my system (vx3 is a slave to the vcal 250 HP).  The set of datapoints that do not give a fault, is added as attachment.  Unfortunately, there are -as far as i can see- no datapoints specific to the vx3.  Basically the same information as for the heatpump is returned. (or do i need to use the option -tx ECUADDR, --ecuaddr ECUADDR sets the default ECU Address different from 0x680 ?)

It is probably different to the case where vx3 works as a stand-alone system.

I could try to connect to the vx3 directly via its own access point. Has anyone tried that yet ?  I'm a bit reluctant to try that.  The Viessmann technician connected his viguide app with the vx3, but after doing so, he had to basically redo the whole configuration, as the vx3 was no longer connected to the vcal as a slave system.  


Next to that,i probably do not tell anything new, there seems to be other datapoints that contain some data.

for example

C:\Users\lauwe\open3e>python3 Open3Eclient.py -d 192.168.0.1 -dev Vx3 -v -r 1844
socket.CMSG_SPACE not available on this platform
Vx3 1844 unknown:len=2 "0004"
closing 0x680 - bye!

C:\Users\lauwe\open3e>python3 Open3Eclient.py -d 192.168.0.1 -dev Vx3 -v -r 2806
socket.CMSG_SPACE not available on this platform
Vx3 2806 unknown:len=2 "0101"
closing 0x680 - bye!

C:\Users\lauwe\open3e>python3 Open3Eclient.py -d 192.168.0.1 -dev Vx3 -v -r 2953
socket.CMSG_SPACE not available on this platform
Vx3 2953 unknown:len=10 "0c0025033e0157046402"
closing 0x680 - bye!

[Open3EdatapointsVx3_nanadoe.txt](https://github.com/open3e/open3e/files/15279168/Open3EdatapointsVx3_nanadoe.txt)

_viel spaß_!
Nanadoe

PS i also tried Linux Mint, but so far no luck yet to access the AP. I blame the Linux driver for the wifi dongle for now.

**Reply by philippoo66** [reply]
hi @Nanadoe !

if vx3 is a slave of vcal, then you address it the wrong way. `closing 0x680 - bye!` says that you talked to the main ecu which is vcal. vx3 as a slave is usually on 0x6a1 or somewhat.

best to do depict_system, but currently there is some problem since a lot of code phrases got replaced in a not working way. we hope to get this fixed this weekend.

you might try an older version, the [experimental branch](https://github.com/open3e/open3e/tree/experimental). With this depict_system and -cnfg should work.

**Reply by philippoo66** [reply]
@Nanadoe we re-established the latest stable version in https://github.com/open3e/open3e/tree/no_pip_20240406

please clone that and run `Open3E_depictSystem.py` (may take 1/4 hour or so). after that with Open3Eclient `-cnfg devices.json` you should be able to address all devices/ECUs in your system accurately.

**Comment by Nanadoe** [comment]
Hi Phil,

Thank you for the quick reply. This worked, indeed. Until now, I did indeed
not get the 'depictSystem' to work

below, an overview of what i did this evening:

I cloned the branch using C:\Users\lauwe>git clone -b no_pip_20240406
https://github.com/open3e/open3e.git, i hope that gives the correct
branch.  I'm really not that familiar with all this :-).

I rann the command depict system

C:\Users\lauwe\open3e>python3 Open3E_depictSystem.py -d 192.168.0.1
socket.CMSG_SPACE not available on this platform
read DID enums ...
2951 DIDs listed.
scan COB-IDs 0x680 to 0x6ff ...
ECU found: 0x680 : HPMUMASTER
but, with the next ECU, there is an error.

[image: afbeelding.png]
![afbeelding](https://github.com/open3e/open3e/assets/167920241/3d70130f-ef51-4fae-b5cf-a257c32316cc)

however, when removing 'raise Exception(e)', the script works fine!

in the next step, there is an error '2024-05-11 22:24:40 [ERROR] UdsClient:
[TimeoutError] : ECU failed to respond in time'

[image: afbeelding.png]
![afbeelding](https://github.com/open3e/open3e/assets/167920241/160b063a-7fd6-45f6-a731-6683512632b2)

what works however, is the following:
C:\Users\lauwe\open3e>python3 Open3Eclient.py -d 192.168.0.1 -tx 0x6e6 -r
256
socket.CMSG_SPACE not available on this platform
0x6e6 {'BusAddress': 0, 'BusType': {'ID': 12, 'Text': 'HEMS_Bus'},
'DeviceProperty': {'ID': 18, 'Text': 'PHOTOVOLTAIC'}, 'DeviceFunction':
{'ID': 0, 'Text': 'NOTHING'}, 'SW-Version': '84.503.2243.4', 'HW-Version':
'0.0.0.0', 'VIN': '0'}

many thanks again for the great work and all the kind help
Nanadoe

**Reply by philippoo66** [reply]
your clone command seems to be all right. 

unfortunately the pictures are not visible to me. With the COB scan timeout is the usual reaction if there is no ECU, any other exception _should_ not arise. I don't have no AP so I can not test - perhaps you have to try arround with these lines:
![grafik](https://github.com/open3e/open3e/assets/122479122/d0ca2a04-c195-4733-b9a1-bcca362d9690)
perhaps with the DoIPClientUDSConnector some more than only closing the client has to be done befor opening the next connection for the next try with the next ECU address (line 65)... 

with the `-tx 0x6e6 -r <did>` you should be able to read every existing datapoint of ECU 6e6h (same should work without the -tx option but using `-r 0x6e6.<did>`. What does not work is the -a (read all).

What may work to find all the DIDs of 6e6 is to limit the COB scan range to 6e6 in lines 41, 42: set both to 0x6e6. That way only one connection gets opened, and as with 0x680 it should work once. You can do this with every 'known' ECU address.

**Comment by Nanadoe** [comment]
Thank you!.  I reloaded the screenshots yesterday.  Maybe they are visible now ?
For clarity, depict system worked by suppressing the exception.  It created a devices.json file and several other files with known dids per device.  However when running Open3Eclient -cnfg devices.json , i got the second exception.  

have a nice evening!
